File: Sorting_algorithms/Python/Insertion_sort.py
# Time:
# Best: O(n)
# Average: O(n ^ 2)
# Worst: O(n ^ 2)

# Space: O(1)


# Approach: 1
def insertionSort(arr):
    n = len(arr)
    for i in range(1, n):
        j = i
        while j > 0 and arr[j] < arr[j - 1]:
            arr[j], arr[j - 1] = arr[j - 1], arr[j]
            j -= 1
    
    return arr


# Approach: 2
def insertionSort(arr, n):
    for idx in range(1, n):
        val, j = arr[idx], idx - 1
        
        while j >= 0 and arr[j] > val:
            arr[j + 1], j = arr[j], j - 1
        
        arr[j + 1] = val
    return arr
        

File: Sorting_algorithms/C/Insertion_sort.c
// Time:
// Best: O(n)
// Average: O(n ^ 2)
// Worst: O(n ^ 2)

// Space: O(1)


// Approach: 1
void insertionSort(int* arr) {
    for(int i = 1; arr[i] != '\0'; i++) {
        int j = i;
        while(j > 0 && arr[j] < arr[j - 1]) {
            arr[j] = arr[j] + arr[j - 1];
            arr[j - 1] = arr[j] - arr[j - 1];
            arr[j] = arr[j] - arr[j - 1];
            j--;
        }
    }
}

// Approach: 2
void insertionSort(int* arr, int len) {
    int val, j;
    
    for(int idx = 1; idx < len; idx++) {
        val = arr[idx];
        j = idx - 1;
        
        while(j >= 0 && arr[j] > val) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = val;
    }
}
